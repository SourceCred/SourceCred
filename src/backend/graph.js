// @flow

export type Address = {
  repositoryName: string,
  pluginName: string,
  id: string,
};

export type Node<T> = {
  address: Address,
  inEdges: Address[],
  outEdges: Address[],
  payload: T,
};

export type Edge<T> = {
  address: Address,
  src: Address,
  dst: Address,
  payload: T,
};

export type Graph = {
  nodes: {[stringAddress: string]: Node<mixed>},
  edges: {[stringAddress: string]: Edge<mixed>},
};

export function addressToString(address: Address) {
  if (address.repositoryName.includes("$")) {
    const escaped = JSON.stringify(address.repositoryName);
    throw new Error(`address.repositoryName must not include "\$": ${escaped}`);
  }
  if (address.pluginName.includes("$")) {
    const escaped = JSON.stringify(address.pluginName);
    throw new Error(`address.pluginName must not include "\$": ${escaped}`);
  }
  if (address.id.includes("$")) {
    const escaped = JSON.stringify(address.id);
    throw new Error(`address.id must not include "\$": ${escaped}`);
  }
  return `${address.repositoryName}\$${address.pluginName}\$${address.id}`;
}

export function stringToAddress(string: string) {
  const parts = string.split("$");
  if (parts.length !== 3) {
    const escaped = JSON.stringify(string);
    throw new Error(`Input should have exactly two \$s: ${escaped}`);
  }
  return {
    repositoryName: parts[0],
    pluginName: parts[1],
    id: parts[2],
  };
}
