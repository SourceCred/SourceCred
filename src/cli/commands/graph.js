// @flow

import {Command, flags} from "@oclif/command";
import mkdirp from "mkdirp";
import os from "os";
import path from "path";

import {pluginNames} from "./common";

const execDependencyGraph = require("../../tools/execDependencyGraph").default;

export default class GraphCommand extends Command {
  static description = "create the contribution graph for a repository";

  static args = [
    {
      name: "repo_owner",
      required: true,
      description: "owner of the GitHub repository for which to fetch data",
    },
    {
      name: "repo_name",
      required: true,
      description: "name of the GitHub repository for which to fetch data",
    },
  ];

  static flags = {
    "output-directory": flags.string({
      short: "o",
      description: "directory into which to store graphs",
    }),
    "github-token": flags.string({
      description:
        "a GitHub API token, as generated at " +
        "https://github.com/settings/tokens/new",
      required: true,
      env: "SOURCECRED_GITHUB_TOKEN",
    }),
  };

  async run() {
    const {
      args: {repo_owner: repoOwner, repo_name: repoName},
      flags: {"github-token": token, "output-directory": rawOutputDirectory},
    } = this.parse(GraphCommand);
    const outputDirectory = (() => {
      if (rawOutputDirectory != null) {
        return rawOutputDirectory;
      }
      const outputDirectory = path.join(os.tmpdir(), "sourcecred", repoName);
      this.log("Using output directory: " + outputDirectory);
      return outputDirectory;
    })();
    graph(outputDirectory, repoOwner, repoName, token);
  }
}

function graph(
  outputDirectory: string,
  repoOwner: string,
  repoName: string,
  token: string
) {
  mkdirp.sync(outputDirectory);
  const tasks = makeTasks(outputDirectory, {repoOwner, repoName, token});
  execDependencyGraph(tasks).then(({success}) => {
    process.exitCode = success ? 0 : 1;
  });
}

function makeTasks(outputDirectory, {repoOwner, repoName, token}) {
  const taskId = (id) => `create-${id}`;
  const graphFilename = (id) => path.join(outputDirectory, `graph-${id}.json`);
  const into = "./src/cli/into.sh";
  return [
    ...pluginNames().map((id) => ({
      id: taskId(id),
      cmd: [
        into,
        graphFilename(id),
        "node",
        "./bin/sourcecred.js",
        "plugin-graph",
        "--plugin",
        id,
        repoOwner,
        repoName,
        "--github-token",
        token,
      ],
      deps: [],
    })),
    {
      id: "combine",
      cmd: [
        into,
        path.join(outputDirectory, "graph.json"),
        "node",
        "./bin/sourcecred.js",
        "combine",
        ...pluginNames().map((id) => graphFilename(id)),
      ],
      deps: pluginNames().map((id) => taskId(id)),
    },
  ];
}
