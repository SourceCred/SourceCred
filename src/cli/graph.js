// @flow

import fs from "fs-extra";
import sortBy from "lodash.sortby";
import stringify from "json-stable-stringify";
import {join as pathJoin} from "path";
import {loadFileWithDefault} from "../util/disk";

import {Ledger} from "../ledger/ledger";
import * as NullUtil from "../util/null";
import {LoggingTaskReporter} from "../util/taskReporter";
import {type ReferenceDetector} from "../core/references/referenceDetector";
import {CascadingReferenceDetector} from "../core/references/cascadingReferenceDetector";
import type {Command} from "./command";
import dedent from "../util/dedent";
import {type InstanceConfig} from "../api/instanceConfig";
import {
  makePluginDir,
  loadInstanceConfig,
  pluginDirectoryContext,
} from "./common";
import {toJSON as weightedGraphToJSON} from "../core/weightedGraph";
import * as pluginId from "../api/pluginId";
import {ensureIdentityExists} from "../ledger/identityProposal";

function die(std, message) {
  std.err("fatal: " + message);
  return 1;
}

const graphCommand: Command = async (args, std) => {
  const taskReporter = new LoggingTaskReporter();
  taskReporter.start("graph");
  const baseDir = process.cwd();
  const config: InstanceConfig = await loadInstanceConfig(baseDir);

  const ledgerPath = pathJoin(baseDir, "data", "ledger.json");
  const ledger = Ledger.parse(
    await loadFileWithDefault(ledgerPath, () => new Ledger().serialize())
  );

  let pluginsToLoad = [];
  if (args.length === 0) {
    pluginsToLoad = config.bundledPlugins.keys();
  } else {
    for (const arg of args) {
      const id = pluginId.fromString(arg);
      if (config.bundledPlugins.has(id)) {
        pluginsToLoad.push(id);
      } else {
        return die(
          std,
          `can't find plugin ${id}; remember to use fully scoped name, as in sourcecred/github`
        );
      }
    }
  }

  const graphOutputPrefix = ["output", "graphs"];

  const rd = await buildReferenceDetector(
    baseDir,
    config,
    taskReporter,
    ledger
  );
  for (const name of pluginsToLoad) {
    const plugin = NullUtil.get(config.bundledPlugins.get(name));
    const task = `${name}: generating graph`;
    taskReporter.start(task);
    const dirContext = pluginDirectoryContext(baseDir, name);
    const graph = await plugin.graph(dirContext, rd, taskReporter);
    const serializedGraph = stringify(weightedGraphToJSON(graph));
    const outputDir = makePluginDir(baseDir, graphOutputPrefix, name);
    const outputPath = pathJoin(outputDir, "graph.json");
    await fs.writeFile(outputPath, serializedGraph);

    const identities = await plugin.identities(dirContext, taskReporter);
    for (const identityProposal of identities) {
      ensureIdentityExists(ledger, identityProposal);
    }
    taskReporter.finish(task);
  }
  await fs.writeFile(ledgerPath, ledger.serialize());
  taskReporter.finish("graph");
  return 0;
};

async function buildReferenceDetector(baseDir, config, taskReporter, ledger) {
  taskReporter.start("reference detector");
  const rds = [];
  for (const [name, plugin] of sortBy(
    [...config.bundledPlugins],
    ([k, _]) => k
  )) {
    const dirContext = pluginDirectoryContext(baseDir, name);
    const task = `reference detector for ${name}`;
    taskReporter.start(task);
    const rd = await plugin.referenceDetector(dirContext, taskReporter);
    rds.push(rd);
    taskReporter.finish(task);
  }
  taskReporter.finish("reference detector");
  rds.push(_hackyIdentityNameReferenceDetector(ledger));
  return new CascadingReferenceDetector(rds);
}

// Hack to support old-school (deprecated) "initiatives files":
// We need to be able to parse references to usernames, e.g. "@yalor", so
// we need a reference detector that will match these to identities in the
// Ledger. This isn't a robust addressing scheme, since identities are re-nameable;
// in v2 the initiatives plugin will be re-written to use identity node addresses instead.
// This hack can be safely deleted once we no longer support initiatives files that refer
// to identities by their names instead of their IDs.
export function _hackyIdentityNameReferenceDetector(
  ledger: Ledger
): ReferenceDetector {
  const usernameToAddress = new Map(
    ledger.accounts().map((a) => [a.identity.name, a.identity.address])
  );
  function addressFromUrl(potentialUsername: string) {
    const prepped = potentialUsername.replace("@", "").toLowerCase();
    return usernameToAddress.get(prepped);
  }
  return {addressFromUrl};
}

export const graphHelp: Command = async (args, std) => {
  std.out(
    dedent`\
      usage: sourcecred graph

      Generate a graph from cached plugin data

      Either 'sourcecred load' must immediately precede this command or
      a cache directory must exist for all plugins specified in sourcecred.json
      `.trimRight()
  );
  return 0;
};

export default graphCommand;
