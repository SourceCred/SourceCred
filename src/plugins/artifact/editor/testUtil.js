// @flow

export function configureEnzyme() {
  const Enzyme = require("enzyme");
  const Adapter = require("enzyme-adapter-react-16");
  Enzyme.configure({adapter: new Adapter()});
}

export function configureAphrodite() {
  const {StyleSheetTestUtils} = require("aphrodite/no-important");
  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });
}
