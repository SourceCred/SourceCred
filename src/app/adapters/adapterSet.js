// @flow

import {Graph, type NodeAddressT, type EdgeAddressT} from "../../core/graph";
import {NodeTrie, EdgeTrie} from "../../core/trie";
import type {Assets} from "../assets";
import type {RepoId} from "../../core/repoId";

import type {
  StaticPluginAdapter,
  DynamicPluginAdapter,
  NodeType,
  EdgeType,
} from "./pluginAdapter";

import {FallbackStaticAdapter} from "./fallbackAdapter";

export class StaticAdapterSet {
  _adapters: $ReadOnlyArray<StaticPluginAdapter>;
  _adapterNodeTrie: NodeTrie<StaticPluginAdapter>;
  _adapterEdgeTrie: EdgeTrie<StaticPluginAdapter>;
  _typeNodeTrie: NodeTrie<NodeType>;
  _typeEdgeTrie: EdgeTrie<EdgeType>;

  constructor(adapters: $ReadOnlyArray<StaticPluginAdapter>): void {
    this._adapters = [new FallbackStaticAdapter(), ...adapters];
    this._adapterNodeTrie = new NodeTrie();
    this._adapterEdgeTrie = new EdgeTrie();
    this._typeNodeTrie = new NodeTrie();
    this._typeEdgeTrie = new EdgeTrie();
    const usedPluginNames = new Set();
    this._adapters.forEach((a) => {
      const name = a.name();
      if (usedPluginNames.has(name)) {
        throw new Error(`Multiple plugins with name "${name}"`);
      }
      usedPluginNames.add(name);
      this._adapterNodeTrie.add(a.nodePrefix(), a);
      this._adapterEdgeTrie.add(a.edgePrefix(), a);
    });
    this.nodeTypes().forEach((t) => this._typeNodeTrie.add(t.prefix, t));
    this.edgeTypes().forEach((t) => this._typeEdgeTrie.add(t.prefix, t));
  }

  adapters(): $ReadOnlyArray<StaticPluginAdapter> {
    return this._adapters;
  }

  nodeTypes(): NodeType[] {
    return [].concat(...this._adapters.map((x) => x.nodeTypes()));
  }

  edgeTypes(): EdgeType[] {
    return [].concat(...this._adapters.map((x) => x.edgeTypes()));
  }

  adapterMatchingNode(x: NodeAddressT): StaticPluginAdapter {
    return this._adapterNodeTrie.getLast(x);
  }

  adapterMatchingEdge(x: EdgeAddressT): StaticPluginAdapter {
    return this._adapterEdgeTrie.getLast(x);
  }

  typeMatchingNode(x: NodeAddressT): NodeType {
    return this._typeNodeTrie.getLast(x);
  }

  typeMatchingEdge(x: EdgeAddressT): EdgeType {
    return this._typeEdgeTrie.getLast(x);
  }

  load(assets: Assets, repoId: RepoId): Promise<DynamicAdapterSet> {
    return Promise.all(this._adapters.map((a) => a.load(assets, repoId))).then(
      (adapters) => new DynamicAdapterSet(this, adapters)
    );
  }
}

export class DynamicAdapterSet {
  _adapters: $ReadOnlyArray<DynamicPluginAdapter>;
  _staticAdapterSet: StaticAdapterSet;
  _adapterNodeTrie: NodeTrie<DynamicPluginAdapter>;
  _adapterEdgeTrie: EdgeTrie<DynamicPluginAdapter>;

  constructor(
    staticAdapterSet: StaticAdapterSet,
    adapters: $ReadOnlyArray<DynamicPluginAdapter>
  ): void {
    this._staticAdapterSet = staticAdapterSet;
    this._adapters = adapters;
    this._adapterNodeTrie = new NodeTrie();
    this._adapterEdgeTrie = new EdgeTrie();
    this._adapters.forEach((a) => {
      this._adapterNodeTrie.add(a.static().nodePrefix(), a);
      this._adapterEdgeTrie.add(a.static().edgePrefix(), a);
    });
  }

  adapterMatchingNode(x: NodeAddressT): DynamicPluginAdapter {
    return this._adapterNodeTrie.getLast(x);
  }

  adapterMatchingEdge(x: EdgeAddressT): DynamicPluginAdapter {
    return this._adapterEdgeTrie.getLast(x);
  }

  adapters(): $ReadOnlyArray<DynamicPluginAdapter> {
    return this._adapters;
  }

  graph(): Graph {
    return Graph.merge(this._adapters.map((x) => x.graph()));
  }

  static() {
    return this._staticAdapterSet;
  }
}
