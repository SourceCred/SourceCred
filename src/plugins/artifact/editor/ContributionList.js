// @flow

import React from "react";

import type {Node} from "@/core/graph";
import {AdapterSet} from "./adapterSet";
import {Graph} from "@/core/graph";

type Props = {
  graph: ?Graph<any, any>,
  adapters: AdapterSet,
};
type State = {
  typeFilter: ?{|
    +pluginName: string,
    +type: string,
  |},
};

export class ContributionList extends React.Component<Props, State> {
  constructor() {
    super();
    this.state = {
      typeFilter: null,
    };
  }

  render() {
    return (
      <div>
        <h2>Contributions</h2>
        {this.renderFilterSelect()}
        {this.renderTable()}
      </div>
    );
  }

  renderFilterSelect() {
    if (this.props.graph == null) {
      return null;
    }
    const graph: Graph<any, any> = this.props.graph;
    const typesByPlugin: {[pluginName: string]: Set<string>} = {};
    graph.nodes().forEach((node) => {
      const adapter = this.props.adapters.getAdapter(node);
      if (adapter == null) {
        return;
      }
      if (!typesByPlugin[adapter.pluginName]) {
        typesByPlugin[adapter.pluginName] = new Set();
      }
      typesByPlugin[adapter.pluginName].add(node.address.type);
    });
    function optionGroup(pluginName: string) {
      const header = (
        <option key={pluginName} disabled style={{fontWeight: "bold"}}>
          {pluginName}
        </option>
      );
      const entries = Array.from(typesByPlugin[pluginName])
        .sort()
        .map((type) => (
          <option key={type} value={JSON.stringify({pluginName, type})}>
            {"\u2003" + type}
          </option>
        ));
      return [header, ...entries];
    }
    return (
      <label>
        Filter by contribution type:{" "}
        <select
          value={JSON.stringify(this.state.typeFilter)}
          onChange={(e) => {
            this.setState({typeFilter: JSON.parse(e.target.value)});
          }}
        >
          <option value={JSON.stringify(null)}>Show all</option>
          {Object.keys(typesByPlugin)
            .sort()
            .map(optionGroup)}
        </select>
      </label>
    );
  }

  renderTable() {
    if (this.props.graph == null) {
      return <div>(no graph)</div>;
    } else {
      const graph: Graph<any, any> = this.props.graph;
      const {typeFilter} = this.state;
      const shouldDisplay: (node: Node<any>) => boolean = typeFilter
        ? (node) => {
            const adapter = this.props.adapters.getAdapter(node);
            return (
              !!adapter &&
              adapter.pluginName === typeFilter.pluginName &&
              node.address.type === typeFilter.type
            );
          }
        : (_) => true;
      return (
        <table>
          <thead>
            <tr>
              <th>Title</th>
              <th>Artifact</th>
              <th>Weight</th>
            </tr>
          </thead>
          <tbody>
            {this.props.graph.nodes().map((node) => {
              if (!shouldDisplay(node)) {
                return null;
              }
              const adapter = this.props.adapters.getAdapter(node);
              if (adapter == null) {
                return (
                  <tr key={JSON.stringify(node.address)}>
                    <td colSpan={3}>
                      <i>unknown</i> (plugin: {node.address.pluginName})
                    </td>
                  </tr>
                );
              } else {
                return (
                  <tr key={JSON.stringify(node.address)}>
                    <td>{adapter.extractTitle(graph, node)}</td>
                    <td>[TODO]</td>
                    <td>[TODO]</td>
                  </tr>
                );
              }
            })}
          </tbody>
        </table>
      );
    }
  }
}
