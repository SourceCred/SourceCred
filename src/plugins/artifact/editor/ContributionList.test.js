// @flow

import type {ReactWrapper} from "enzyme";
import React from "react";
import {shallow} from "enzyme";
import enzymeToJSON from "enzyme-to-json";

import type {Address} from "../../../core/address";
import type {Node} from "../../../core/graph";
import type {PluginAdapter} from "./pluginAdapter";
import {AdapterSet} from "./adapterSet";
import {ContributionList} from "./ContributionList";
import {Graph} from "../../../core/graph";

require("./testUtil").configureAphrodite();
require("./testUtil").configureEnzyme();

function createTestData(): * {
  type PayloadA = number;
  type PayloadB = boolean;
  type PayloadC = string;
  type NodePayload = PayloadA | PayloadB | PayloadC;
  type EdgePayload = null;

  const PLUGIN_A = "sourcecred/example-plugin-a";
  const PLUGIN_B = "sourcecred/example-plugin-b";
  const PLUGIN_C = "sourcecred/example-plugin-c";

  function makeAddress(
    pluginName: typeof PLUGIN_A | typeof PLUGIN_B | typeof PLUGIN_C,
    type: string,
    id: string
  ): Address {
    return {
      pluginName,
      id,
      type,
    };
  }

  const nodeA1 = () => ({
    address: makeAddress(PLUGIN_A, "small", "one"),
    payload: (111: PayloadA),
  });
  const nodeA2 = () => ({
    address: makeAddress(PLUGIN_A, "small", "two"),
    payload: (234: PayloadA),
  });
  const nodeA3 = () => ({
    address: makeAddress(PLUGIN_A, "big", "three"),
    payload: (616: PayloadA),
  });
  const nodeB4 = () => ({
    address: makeAddress(PLUGIN_B, "very true", "four"),
    payload: (true: PayloadB),
  });
  const nodeC5 = () => ({
    address: makeAddress(PLUGIN_C, "ctype", "five"),
    payload: ("I have no adapter :-(": PayloadC),
  });
  const edgeA1A2 = () => ({
    address: makeAddress(PLUGIN_A, "atype", "one-to-two"),
    payload: null,
    src: nodeA1().address,
    dst: nodeA2().address,
  });
  const edgeB4A3 = () => ({
    address: makeAddress(PLUGIN_C, "ctype", "four-to-three"),
    payload: null,
    src: nodeB4().address,
    dst: nodeA3().address,
  });

  const graph: () => Graph<NodePayload, EdgePayload> = () =>
    new Graph()
      .addNode(nodeA1())
      .addNode(nodeA2())
      .addNode(nodeA3())
      .addNode(nodeB4())
      .addNode(nodeC5())
      .addEdge(edgeA1A2())
      .addEdge(edgeB4A3());

  const adapterA: () => PluginAdapter<PayloadA> = () => ({
    pluginName: PLUGIN_A,
    renderer: class RendererA extends React.Component<{
      graph: Graph<any, any>,
      node: Node<PayloadA>,
    }> {
      render() {
        const {graph, node} = this.props;
        const neighborCount = graph.neighborhood(node.address, {
          direction: "OUT",
        }).length;
        return (
          <span>
            <tt>{node.address.id}</tt> has neighbor count{" "}
            <strong>{neighborCount}</strong>
          </span>
        );
      }
    },
    extractTitle(graph: Graph<NodePayload, EdgePayload>, node: Node<PayloadA>) {
      return `the number ${String(node.payload)}`;
    },
  });

  const adapterB: () => PluginAdapter<PayloadB> = () => ({
    pluginName: PLUGIN_B,
    renderer: class RendererB extends React.Component<{
      graph: Graph<any, any>,
      node: Node<PayloadB>,
    }> {
      render() {
        const {node} = this.props;
        return (
          <span>
            Node <em>{node.address.id}</em>: <strong>{node.payload}</strong>
          </span>
        );
      }
    },
    extractTitle(graph: Graph<NodePayload, EdgePayload>, node: Node<PayloadB>) {
      return String(node.payload).toUpperCase() + "!";
    },
  });

  const adapters: () => AdapterSet = () => {
    const result = new AdapterSet();
    result.addAdapter(adapterA());
    result.addAdapter(adapterB());
    return result;
  };

  return {
    PLUGIN_A,
    PLUGIN_B,
    PLUGIN_C,
    nodeA1,
    nodeA2,
    nodeA3,
    nodeB4,
    nodeC5,
    edgeA1A2,
    edgeB4A3,
    graph,
    adapterA,
    adapterB,
    adapters,
  };
}

describe("ContributionList", () => {
  // Render a contribution list with the above test data.
  function render() {
    const data = createTestData();
    const result = shallow(
      <ContributionList graph={data.graph()} adapters={data.adapters()} />
    );
    return result;
  }

  // Select the unique <option> whose text matches the given patern.
  function simulateSelect(container: ReactWrapper, pattern: RegExp): void {
    const targetOption = container
      .find("option")
      .filterWhere((x) => pattern.test(x.text()));
    expect(targetOption).toHaveLength(1);
    container
      .find("select")
      .simulate("change", {target: {value: targetOption.prop("value")}});
  }

  it("renders some test data in the default state", () => {
    const result = render();
    expect(enzymeToJSON(result)).toMatchSnapshot();
  });

  it("updates the node table when a filter is selected", () => {
    const result = render();
    simulateSelect(result, /small/);
    expect(enzymeToJSON(result)).toMatchSnapshot();
  });

  it("resets the node table when a filter is deselected", () => {
    const result = render();
    const originalHtml = result.html();
    simulateSelect(result, /big/);
    const intermediateHtml = result.html();
    simulateSelect(result, /Show all/);
    const finalHtml = result.html();
    expect(finalHtml).toEqual(originalHtml);
    expect(finalHtml).not.toEqual(intermediateHtml);
  });
});
