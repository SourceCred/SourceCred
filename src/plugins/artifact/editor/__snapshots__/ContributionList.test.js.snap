// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContributionList renders some test data in the default state 1`] = `
<div>
  <h2>
    Contributions
  </h2>
  <label>
    Filter by contribution type:
     
    <select
      onChange={[Function]}
      value="null"
    >
      <option
        value="null"
      >
        Show all
      </option>
      <option
        disabled={true}
        key="sourcecred/example-plugin-a"
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        sourcecred/example-plugin-a
      </option>
      <option
        key="big"
        value="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"type\\":\\"big\\"}"
      >
         big
      </option>
      <option
        key="small"
        value="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"type\\":\\"small\\"}"
      >
         small
      </option>
      <option
        disabled={true}
        key="sourcecred/example-plugin-b"
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        sourcecred/example-plugin-b
      </option>
      <option
        key="very true"
        value="{\\"pluginName\\":\\"sourcecred/example-plugin-b\\",\\"type\\":\\"very true\\"}"
      >
         very true
      </option>
    </select>
  </label>
  <table>
    <thead>
      <tr>
        <th>
          Title
        </th>
        <th>
          Artifact
        </th>
        <th>
          Weight
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"id\\":\\"one\\",\\"type\\":\\"small\\"}"
      >
        <td>
          the number 111
        </td>
        <td>
          [TODO]
        </td>
        <td>
          [TODO]
        </td>
      </tr>
      <tr
        key="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"id\\":\\"two\\",\\"type\\":\\"small\\"}"
      >
        <td>
          the number 234
        </td>
        <td>
          [TODO]
        </td>
        <td>
          [TODO]
        </td>
      </tr>
      <tr
        key="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"id\\":\\"three\\",\\"type\\":\\"big\\"}"
      >
        <td>
          the number 616
        </td>
        <td>
          [TODO]
        </td>
        <td>
          [TODO]
        </td>
      </tr>
      <tr
        key="{\\"pluginName\\":\\"sourcecred/example-plugin-b\\",\\"id\\":\\"four\\",\\"type\\":\\"very true\\"}"
      >
        <td>
          TRUE!
        </td>
        <td>
          [TODO]
        </td>
        <td>
          [TODO]
        </td>
      </tr>
      <tr
        key="{\\"pluginName\\":\\"sourcecred/example-plugin-c\\",\\"id\\":\\"five\\",\\"type\\":\\"ctype\\"}"
      >
        <td
          colSpan={3}
        >
          <i>
            unknown
          </i>
           (plugin: 
          sourcecred/example-plugin-c
          )
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ContributionList updates the node table when a filter is selected 1`] = `
<div>
  <h2>
    Contributions
  </h2>
  <label>
    Filter by contribution type:
     
    <select
      onChange={[Function]}
      value="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"type\\":\\"small\\"}"
    >
      <option
        value="null"
      >
        Show all
      </option>
      <option
        disabled={true}
        key="sourcecred/example-plugin-a"
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        sourcecred/example-plugin-a
      </option>
      <option
        key="big"
        value="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"type\\":\\"big\\"}"
      >
         big
      </option>
      <option
        key="small"
        value="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"type\\":\\"small\\"}"
      >
         small
      </option>
      <option
        disabled={true}
        key="sourcecred/example-plugin-b"
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        sourcecred/example-plugin-b
      </option>
      <option
        key="very true"
        value="{\\"pluginName\\":\\"sourcecred/example-plugin-b\\",\\"type\\":\\"very true\\"}"
      >
         very true
      </option>
    </select>
  </label>
  <table>
    <thead>
      <tr>
        <th>
          Title
        </th>
        <th>
          Artifact
        </th>
        <th>
          Weight
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"id\\":\\"one\\",\\"type\\":\\"small\\"}"
      >
        <td>
          the number 111
        </td>
        <td>
          [TODO]
        </td>
        <td>
          [TODO]
        </td>
      </tr>
      <tr
        key="{\\"pluginName\\":\\"sourcecred/example-plugin-a\\",\\"id\\":\\"two\\",\\"type\\":\\"small\\"}"
      >
        <td>
          the number 234
        </td>
        <td>
          [TODO]
        </td>
        <td>
          [TODO]
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
